### ISSUE TITLE
WebDAV uploaded files always ended empty on server

### ISSUE THREAD

**Issue by Cabbagec** [drakkan/sftpgo#239]
Hi, I appreciate sftpgo a lot for its compatibility, and thanks a lot for your great contribution! But recently, I found that my uploads via webdav all became empty files. I don't think there was some problem with file or directory permissions, because the files were created just fine. It was just all of them become 0 sized after successful uploading. Logs are captured as below while I was trying to upload multiple times.

I was using sftpgo 1.2.2 on docker with image of `ghcr.io/drakkan/sftpgo:edge`, and run this container via command:
```
docker run --name sftpgo \
--cap-add NET_BIND_SERVICE \
--sysctl net.ipv4.ip_unprivileged_port_start=0 \
-p 127.0.0.1:10125:8080 \
-p 127.0.0.1:10126:8081 \
-p 2022:2022 \
-p 21:21 \
-p 20:20 \
-p 50000-50100:50000-50100 \
-e SFTPGO_FTPD__BIND_PORT=21 \
-e SFTPGO_WEBDAVD__BIND_PORT=8081 \
-e SFTPGO_SFTPD__TLS_MODE=0 \
--mount type=bind,source=/var/www/sftpgo/data,target=/srv/sftpgo \
--mount type=bind,source=/var/www/sftpgo,target=/var/lib/sftpgo \
-d --restart always ghcr.io/drakkan/sftpgo:edge
```

And I was using Windows 10 20H2 explorer. I mounted this webdav service as a network drive.

These logs just said something like `uploaded file size 0 stat error: <nil>`, `uploaded file size 23414 stat error: <nil>`. If I can provide something else that can be of help, please do let me know. Thank you very much!

```
{"level":"debug","time":"2020-12-05T16:23:23.369","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5og","message":"connection removed, num open connections: 0"}
{"level":"debug","time":"2020-12-05T16:23:23.799","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"connection added, num open connections: 1"}
{"level":"debug","time":"2020-12-05T16:23:23.800","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"transfer added, id: 1, active transfers: 1"}
{"level":"info","time":"2020-12-05T16:23:23.800","sender":"Download","elapsed_ms":0,"size_bytes":0,"username":"user_webdav","file_path":"/srv/sftpgo/data/user_webdav/autoproxy.min.bak","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","protocol":"DAV"}
{"level":"debug","time":"2020-12-05T16:23:23.800","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"transfer removed, id: 1 active transfers: 0"}
{"level":"info","time":"2020-12-05T16:23:23.800","sender":"webdavd","elapsed_ms":0,"method":"HEAD","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7ur99tfgdqlcv5p0","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
{"level":"debug","time":"2020-12-05T16:23:23.800","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"connection removed, num open connections: 0"}
{"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"connection added, num open connections: 1"}
{"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"transfer added, id: 1, active transfers: 1"}
{"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"uploaded file size 23414 stat error: <nil>"}
{"level":"info","time":"2020-12-05T16:23:24.453","sender":"Upload","elapsed_ms":0,"size_bytes":23414,"username":"user_webdav","file_path":"/srv/sftpgo/data/user_webdav/autoproxy.min.bak","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","protocol":"DAV"}
{"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"transfer removed, id: 1 active transfers: 0"}
{"level":"info","time":"2020-12-05T16:23:24.453","sender":"webdavd","elapsed_ms":0,"method":"PUT","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7v399tfgdqlcv5pg","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
{"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"connection removed, num open connections: 0"}
{"level":"debug","time":"2020-12-05T16:23:24.667","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"connection added, num open connections: 1"}
{"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"transfer added, id: 1, active transfers: 1"}
{"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"uploaded file size 0 stat error: <nil>"}
{"level":"info","time":"2020-12-05T16:23:24.668","sender":"Upload","elapsed_ms":0,"size_bytes":0,"username":"user_webdav","file_path":"/srv/sftpgo/data/user_webdav/autoproxy.min.bak","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","protocol":"DAV"}
{"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"transfer removed, id: 1 active transfers: 0"}
{"level":"info","time":"2020-12-05T16:23:24.668","sender":"webdavd","elapsed_ms":1,"method":"PROPPATCH","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7v399tfgdqlcv5q0","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
{"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"connection removed, num open connections: 0"}
{"level":"debug","time":"2020-12-05T16:23:24.884","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5qg","message":"connection added, num open connections: 1"}
{"level":"info","time":"2020-12-05T16:23:24.884","sender":"webdavd","elapsed_ms":0,"method":"UNLOCK","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7v399tfgdqlcv5qg","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
```

**Comment by drakkan** [comment]
> Hi, I appreciate sftpgo a lot for its compatibility, and thanks a lot for your great contribution! 

thank you

> But recently, I found that my uploads via webdav all became empty files. 

do you mean that they worked before?

> I don't think there was some problem with file or directory permissions, because the files were created just fine. It was just all of them become 0 sized after successful uploading. Logs are captured as below while I was trying to upload multiple times.
> 
> I was using sftpgo 1.2.2 on docker with image of `ghcr.io/drakkan/sftpgo:edge`, and run this container via command:
> 

`sftpgo:edge` is a floating tag for the latest commit, can you provide the exact version? `sftpgo -v` or the version printed in the logs at startup

> ```
> docker run --name sftpgo \
> --cap-add NET_BIND_SERVICE \
> --sysctl net.ipv4.ip_unprivileged_port_start=0 \
> -p 127.0.0.1:10125:8080 \
> -p 127.0.0.1:10126:8081 \
> -p 2022:2022 \
> -p 21:21 \
> -p 20:20 \
> -p 50000-50100:50000-50100 \
> -e SFTPGO_FTPD__BIND_PORT=21 \
> -e SFTPGO_WEBDAVD__BIND_PORT=8081 \
> -e SFTPGO_SFTPD__TLS_MODE=0 \
> --mount type=bind,source=/var/www/sftpgo/data,target=/srv/sftpgo \
> --mount type=bind,source=/var/www/sftpgo,target=/var/lib/sftpgo \
> -d --restart always ghcr.io/drakkan/sftpgo:edge
> ```
> 
> And I was using Windows 10 20H2 explorer. I mounted this webdav service as a network drive.
> 
> These logs just said something like `uploaded file size 0 stat error: <nil>`, `uploaded file size 23414 stat error: <nil>`. If I can provide something else that can be of help, please do let me know. Thank you very much!
> 
> ```
> {"level":"debug","time":"2020-12-05T16:23:23.369","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5og","message":"connection removed, num open connections: 0"}
> {"level":"debug","time":"2020-12-05T16:23:23.799","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"connection added, num open connections: 1"}
> {"level":"debug","time":"2020-12-05T16:23:23.800","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"transfer added, id: 1, active transfers: 1"}
> {"level":"info","time":"2020-12-05T16:23:23.800","sender":"Download","elapsed_ms":0,"size_bytes":0,"username":"user_webdav","file_path":"/srv/sftpgo/data/user_webdav/autoproxy.min.bak","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","protocol":"DAV"}
> {"level":"debug","time":"2020-12-05T16:23:23.800","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"transfer removed, id: 1 active transfers: 0"}
> {"level":"info","time":"2020-12-05T16:23:23.800","sender":"webdavd","elapsed_ms":0,"method":"HEAD","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7ur99tfgdqlcv5p0","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
> {"level":"debug","time":"2020-12-05T16:23:23.800","sender":"DAV","connection_id":"DAV_bv5r7ur99tfgdqlcv5p0","message":"connection removed, num open connections: 0"}
> {"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"connection added, num open connections: 1"}
> {"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"transfer added, id: 1, active transfers: 1"}
> {"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"uploaded file size 23414 stat error: <nil>"}
> {"level":"info","time":"2020-12-05T16:23:24.453","sender":"Upload","elapsed_ms":0,"size_bytes":23414,"username":"user_webdav","file_path":"/srv/sftpgo/data/user_webdav/autoproxy.min.bak","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","protocol":"DAV"}
> {"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"transfer removed, id: 1 active transfers: 0"}
> {"level":"info","time":"2020-12-05T16:23:24.453","sender":"webdavd","elapsed_ms":0,"method":"PUT","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7v399tfgdqlcv5pg","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
> {"level":"debug","time":"2020-12-05T16:23:24.453","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5pg","message":"connection removed, num open connections: 0"}
> {"level":"debug","time":"2020-12-05T16:23:24.667","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"connection added, num open connections: 1"}
> {"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"transfer added, id: 1, active transfers: 1"}
> {"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"uploaded file size 0 stat error: <nil>"}
> {"level":"info","time":"2020-12-05T16:23:24.668","sender":"Upload","elapsed_ms":0,"size_bytes":0,"username":"user_webdav","file_path":"/srv/sftpgo/data/user_webdav/autoproxy.min.bak","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","protocol":"DAV"}
> {"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"transfer removed, id: 1 active transfers: 0"}
> {"level":"info","time":"2020-12-05T16:23:24.668","sender":"webdavd","elapsed_ms":1,"method":"PROPPATCH","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7v399tfgdqlcv5q0","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
> {"level":"debug","time":"2020-12-05T16:23:24.668","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5q0","message":"connection removed, num open connections: 0"}
> {"level":"debug","time":"2020-12-05T16:23:24.884","sender":"DAV","connection_id":"DAV_bv5r7v399tfgdqlcv5qg","message":"connection added, num open connections: 1"}
> {"level":"info","time":"2020-12-05T16:23:24.884","sender":"webdavd","elapsed_ms":0,"method":"UNLOCK","proto":"HTTP/1.0","remote_addr":"<actual.ip.replaced>","request_id":"bv5r7v399tfgdqlcv5qg","uri":"http://<actual.domain.replaced>/user_webdav/autoproxy.min.bak","user_agent":"Microsoft-WebDAV-MiniRedir/10.0.19042"}
> ```

from the logs it appears that a file is first uploaded and then replaced with one of zero bytes size. A wireshark network dump could help here, can you provide it? 

I tested recently windows explorer, see [here](https://github.com/drakkan/sftpgo/issues/207#issuecomment-721741248) and it worked after I fixed the referenced issue, maybe there is another similar bug

Thank you

**Comment by drakkan** [comment]
Hi,

I was able to reproduce and fix the issue. I'll push the fix tomorrow after some other tests, thank you for reporting

**Comment by Cabbagec** [comment]
Hi!
It did work before (probably on macOS finder, I don't remember exactly). 

I think it happened after upgrading to this version below, which I'm using now.
```
SFTPGo 1.2.2-dev-4a88ea5c-2020-12-05T12:51:54Z +metrics +awskms +gcpkms +vaultkms +azblob +gcs +s3 +bolt +mysql +pgsql +sqlite +portable
```

>  Hi,
>
> I was able to reproduce and fix the issue. I'll push the fix tomorroww after some other tests, thank you for reporting

Well, I'll provide traffic dump on it if you ever need it. And, thank you for your awesome work!

**Comment by drakkan** [comment]
Hi, 

this issue happen in any SFTPGo version if the client issues a `PROPPATCH` request after a `PUT`.
If you look at the logs provided in the other issue, explorer used a different command sequence at that time, anyway this is a SFTPGo bug and it is now covered by a specific test case so it should not happen anymore.
Can you please confirm that it is fixed in `1.2.2-dev-034d898`? Thank you

**Comment by Cabbagec** [comment]
> Can you please confirm that it is fixed in 1.2.2-dev-034d898? Thank you

Yes, problem solved, you are awesome!

**Comment by drakkan** [comment]
> > Can you please confirm that it is fixed in 1.2.2-dev-034d898? Thank you
> 
> Yes, problem solved, you are awesome!

Thank you for confirming. If you want to ensure that the project stays healthy and well maintained as it is now, please consider a [sponsorship](https://github.com/sponsors/drakkan)

**Comment by drakkan** [comment]
Thank you for your sponsorship, much appreciated!
